corehq/apps/es/README.rst

```diff
@@ -1,6 +1,174 @@
+
+When modifying mappings for an existing index, it can be useful to compare the
+new mapping (as defined in code) to the live index mappings in Elasticsearch on
+a CommCare HQ deployment. This is possible by dumping the mappings of interest
+into local files and comparing them with a diff utility. The
+``print_elastic_mappings`` Django manage command makes this process relatively
+easy. Minimally, this can be accomplished in as few as three steps:
+
+1. Export the local code mapping into a new file.
+2. Export the mappings from a deployed environment into a local file.
+3. Compare the two files.
+
+In practice, this might look like the following example:
+
+.. code-block:: shell
+
+   ./manage.py print_elastic_mappings sms --no-names > ./sms-in-code.py
+   cchq <env> django-manage print_elastic_mappings smslogs_2020-01-28:sms --no-names > ./sms-live.py
+   diff -u ./sms-live.py ./sms-in-code.py
+
+
+Elastic Index Tuning Configurations
+'''''''''''''''''''''''''''''''''''
+
+CommCare HQ provides a mechanism for individual deployments (environments) to
+tune the performance characteristics of their Elasticsearch indexes via Django
+settings. This mechanism can be used by defining an ``ES_SETTINGS`` dictionary
+in ``localsettings.py`` (or by configuring the requisite Elasticsearch
+parameters in a `CommCare Cloud environment`_). Tuning parameters can be
+specified in one of two ways:
+
+1. **"default"**: configures the tuning settings for *all* indexes in the
+   environment.
+2. **index identifier**: configures the tuning settings for *a specific* index
+   in the environment -- these settings take precedence over "default" settings.
+
+For example, if an environment wishes to explicitly configure the "case_search"
+index with six shards, and all others with only three, the configuration could
+be specified in ``localsettings.py`` as:
+
+.. code-block:: python
+
+   ES_SETTINGS = {
+       "default": {"number_of_shards": 3},
+       "case_search": {"number_of_shards": 6},
+   }
+
+Configuring a tuning setting with the special value ``None`` will result in that
+configuration item being reset to the Elasticsearch cluster default (unless
+superseded by another setting with higher precedence). Refer to
+`corehq/app/es/index/settings.py`_ file for the full details regarding what
+items (index and tunning settings values) are configurable, as well as what
+default tuning settings will be used when not customized by the environment.
+
+**Important note**: These Elasticsearch index tuning settings are not "live".
+That is: changing their values on a deployed environment will not have any
+immediate affect on live indexes in Elasticsearch. Instead, these values are
+only ever used when an index is created (for example, during a fresh CommCare HQ
+installation or when an existing index is reindexed into a new one). This means
+that making new values become "live" requires a update to the CommCare HQ
+codebase.
```

Comment:
Can this be a bit more specific about what kind of update would be required to do that? I think it would involve an index migration, but having that stated explicitly here might be valuable.

Reply:
See fe7b237e490